include/envoy/stats/timespan.h

```diff
@@ -24,37 +24,31 @@ class CompletableTimespan {
 };
 
 /**
- * An individual timespan that flushes its measured value in time unit (e.g
- * std::chrono::milliseconds). The initial time is captured on construction. A timespan must be
+ * An individual timespan that flushes its measured value to the histogram on completion.
+ * The start time is captured on construction. The timespan must be
  * completed via complete() for it to be stored. If the timespan is deleted this will be treated as
- * a cancellation.
+ * a cancellation. The target histogram must represent a quantity of time.
  */
-template <class TimeUnit> class TimespanWithUnit : public CompletableTimespan {
+class HistogramCompletableTimespan : public CompletableTimespan {
 public:
-  TimespanWithUnit(Histogram& histogram, TimeSource& time_source)
+  HistogramCompletableTimespan(Histogram& histogram, TimeSource& time_source)
       : time_source_(time_source), histogram_(histogram), start_(time_source.monotonicTime()) {}
 
-  /**
-   * Complete the timespan and send the time to the histogram.
-   */
-  void complete() override { histogram_.recordValue(getRawDuration().count()); }
-
   /**
    * Get duration in the time unit since the creation of the span.
    */
-  TimeUnit getRawDuration() {
+  template <class TimeUnit> TimeUnit getRawDuration() {
     return std::chrono::duration_cast<TimeUnit>(time_source_.monotonicTime() - start_);
   }
 
-private:
+protected:
```

Comment:
leave this private: and move histogram_ into the impl. Actually I'd probably move time_source_ and start_ into the impl as well, otherwise this is not really an interface, but maybe it's not worth it, especially as this was the existing condition.

Reply:
Moved everything and added a virtual `elapsedMs` method to the interface (since I could not use templates), does this look better?